aten/src/THC/THCOffsetInfo.cuh

```diff
@@ -0,0 +1,274 @@
+
+#include "THCIntegerDivider.cuh"
+#include "THCTensorInfo.cuh"
+
+// A faster implementation of IndexToOffset that pre-computes the increments to
+// the indices along each dimension.
+//
+// Consider a kernel with the following loop:
+//
+//      for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+//           linearIndex < totalElements;
+//           linearIndex += gridDim.x * blockDim.x) { ... }
+//
+// Note that the increment value (gridDim.x * blockDim.x) is fixed and known
+// before the kernel starts.  Hence, we can pre-compute how the coordinates
+// change.  For example, assume that the increment is 1024 and the tensor has
+// size [5, 70, 10] and strides [2000, 20, 1].  During the execution of a
+// particular thread, linearIndex may change from 1205 to 2229:
+//
+//      Before: linear index 1205 = 1 * 70 * 10 + 50 * 10 + 5
+//              coordinate (1, 50, 5)
+//              offset = 1 * 2000 + 50 * 20 + 5 * 1 = 3005
+//
+//      Linear index increment = 1024 = 1 * 70 * 10 + 32 * 10 + 4
+//      Coordinate increment = (1, 32, 4)
+//
+//      After: linear index 2029
+//             coordinate (1, 50, 5) + (1, 32, 4) = (2, 82, 9)
+//                        = (3, 12, 9)   // Handle carry over.
+//             offset = 3 * 2000 + 12 * 20 + 9 * 1 = 6249
+//
+// Thus, by pre-computing "coordinate increment" (1, 32, 4), we can compute the
+// next coordinate and offset without costly division.  We also pre-compute the
+// following values, for each dimension k (0 <= k < Dims):
+//
+//      increments[k] == (increment to coordinate #k at each step)
+//      scaledIncrs[k] == increments[k] * strides[k]
+//      carryDelta[k] == strides[k - 1] - (size[k] * strides[k])  (for k > 0)
+//
+// (carryDelta[k] is the change of offset when the addition to dimension #k
+// "carries over" to #k-1.  In general, carryDelta can be "negative" even though
+// IndexType is unsigned: it is always added to the offset, so the result will
+// be correct.)
+//
+// So, the example shown above will be stored as:
+//
+//      increments  = (1, 32, 4)
+//      scaledIncrs = (2000, 640, 4)
+//      carryDelta  = (_, 600, 10)
+//
+// Finally, we can also optimize the initial coordinate computation by
+// pre-computing "magic dividers", thus replacing the initial div/mod operations
+// by multiplication.  (See IntDivider for details.)
+
+// Helper function that increments 'indices' and returns the updated offset.
+template <typename IndexType>
+__host__ __device__ __forceinline__
+IndexType incrementIdx(int dims, IndexType offset,
+                       const IntDivider<IndexType> sizes[],
+                       const IndexType increments[],
+                       const IndexType scaledIncrs[],
+                       const IndexType carryDelta[],
+                       IndexType indices[])
+{
+  bool carry = false;
+
+  for (int i = dims - 1; i > 0; --i) {
+    IndexType index = indices[i] + increments[i] + (IndexType) carry;
```

Comment:
I'm not convinced that maintaining the increment this way is faster than it was before, or especially if you simplify the integer div/mod as above and just calculate the index each time.

This is a lot of instructions to go through, plus the predicate at line 81. I have to imagine that this will bloat out the register count as well, as you're maintaining per-dimension increment state.

The kernels are not as easy to read as before either.

Can you try this code without adding these iterator operators, and have it as before, just by modifying the linear index -> offset code?

Reply:
Done.